server/src/main/java/org/elasticsearch/inference/UnifiedCompletionRequest.java

```diff
@@ -0,0 +1,484 @@
+        }
+
+        public ToolChoiceObject(StreamInput in) throws IOException {
+            this(in.readString(), new FunctionField(in));
+        }
+
+        @Override
+        public void writeTo(StreamOutput out) throws IOException {
+            out.writeString(type);
+            function.writeTo(out);
+        }
+
+        @Override
+        public String getWriteableName() {
+            return NAME;
+        }
+
+        public record FunctionField(String name) implements Writeable {
+            static final ConstructingObjectParser<FunctionField, Void> PARSER = new ConstructingObjectParser<>(
+                "tool_choice_function_field",
+                args -> new FunctionField((String) args[0])
+            );
+
+            static {
+                PARSER.declareString(constructorArg(), new ParseField("name"));
+            }
+
+            public FunctionField(StreamInput in) throws IOException {
+                this(in.readString());
+            }
+
+            @Override
+            public void writeTo(StreamOutput out) throws IOException {
+                out.writeString(name);
+            }
+        }
+    }
+
+    public record ToolChoiceString(String value) implements ToolChoice, NamedWriteable {
+        public static final String NAME = "tool_choice_string";
+
+        public static ToolChoiceString of(XContentParser parser) throws IOException {
+            var content = parser.text();
+            return new ToolChoiceString(content);
+        }
+
+        public ToolChoiceString(StreamInput in) throws IOException {
+            this(in.readString());
+        }
+
+        @Override
+        public void writeTo(StreamOutput out) throws IOException {
+            out.writeString(value);
+        }
+
+        @Override
+        public String getWriteableName() {
+            return NAME;
+        }
+    }
+
+    public record Tool(String type, FunctionField function) implements Writeable {
+
+        static final ConstructingObjectParser<Tool, Void> PARSER = new ConstructingObjectParser<>(
+            Tool.class.getSimpleName(),
+            args -> new Tool((String) args[0], (FunctionField) args[1])
+        );
+
+        static {
+            PARSER.declareString(constructorArg(), new ParseField("type"));
+            PARSER.declareObject(constructorArg(), FunctionField.PARSER::apply, new ParseField("function"));
+        }
+
+        public Tool(StreamInput in) throws IOException {
+            this(in.readString(), new FunctionField(in));
+        }
+
+        @Override
+        public void writeTo(StreamOutput out) throws IOException {
+            out.writeString(type);
+            function.writeTo(out);
+        }
+
+        public record FunctionField(
+            @Nullable String description,
+            String name,
+            @Nullable Map<String, Object> parameters, // TODO can we parse this as a string?
```

Comment:
@jonathan-buttner, I'm wondering if there are any alternative ways to avoid parsing this as a map?